servant/src/Servant/API/Authentication.hs

```diff
@@ -0,0 +1,41 @@
+{-# LANGUAGE DataKinds          #-}
+{-# LANGUAGE DeriveDataTypeable #-}
+{-# LANGUAGE KindSignatures     #-}
+{-# LANGUAGE PolyKinds          #-}
+{-# LANGUAGE TypeFamilies       #-}
+{-# OPTIONS_HADDOCK not-home    #-}
+module Servant.API.Authentication
+( AuthPolicy (..)
+, AuthProtect (..)
+, AuthProtected (..)
+, BasicAuth (..)
+, JWTAuth (..)
+) where
+
+
+import           Data.ByteString (ByteString)
+import           Data.Typeable   (Typeable)
+import           GHC.TypeLits    (Symbol)
+import           Data.Text       (Text)
+
+-- | we can be either Strict or Lax.
+-- Strict: all handlers under 'AuthProtect' take a 'usr' argument.
+--         when auth fails, we call user-supplied handlers to respond.
+-- Lax: all handlers under 'AuthProtect' take a 'Maybe usr' argument.
+--      when auth fails, we call the handlers with 'Nothing'.
+data AuthPolicy = Strict | Lax
+
+-- | the combinator to be used in API types
+data AuthProtect authdata usr (policy :: AuthPolicy)
+
+-- | what we'll ask user to provide at the server-level when we see a
+-- 'AuthProtect' combinator in an API type
+data family AuthProtected authdata usr subserver :: AuthPolicy -> *
```

Comment:
I've been on the sidelines on this one, so apologies in advance for having to ask, but can someone explain exactly why it is that we are using a data family here, rather than a class with an associated type synonym as we usually do?